### PR TITLE
Migrate system socket fields metricset to ECS

### PR DESCRIPTION
**Pull request by jsoriano**

### CODE REVIEW

File: metricbeat/helper/socket/listeners.go [hunk 1]
```diff
@@ -36,10 +36,18 @@ const (
 	Listening
 )
 
+// Names for the direction of a connection
+const (
+	// XXX: Rephrase this to inbound/outbound to match ECS recommendations?
+	IncomingName  = "incoming"
+	OutgoingName  = "outgoing"
+	ListeningName = "listening"
```
Comment (jsoriano) [inline metricbeat/helper/socket/listeners.go]: Should we change to the names recommended by ECS?

We could keep old and new ones in the old and new fields.

File: metricbeat/module/system/socket/socket.go [hunk 2]
```diff
@@ -258,54 +263,107 @@ func newConnection(diag *linux.InetDiagMsg) *connection {
...
 		}
+		metricSetFields["remote"] = remote
+
+		remoteGroup, ok := remoteHostInfoGroup[c.Family.String()]
+		if ok {
+			fields[remoteGroup] = common.MapStr{
+				"ip":   c.RemoteIP.String(),
+				"port": c.RemotePort,
+			}
+		}
```
Comment (webmat) [inline metricbeat/module/system/socket/socket.go]: This made me think to outline the exact mapping I would like to see, in terms of address/ip for the 3 possible states we can observe.

Also made me realize that ECS currently doesn't support sockets that well (noted for later) :-)

**listening**

- local.ip => server.ip
- local.port => server.port

**incoming**

- local.ip => server.ip
- local.port => server.port
- remote.ip => client.ip
- remote.port => client.port

**outgoing**

- local.ip => client.ip
- local.port => client.port
- remote.ip => server.ip
- remote.port => server.port

File: metricbeat/module/system/socket/_meta/data.json [hunk 3]
```diff
@@ -12,27 +12,40 @@
...
-            "user": {
-                "id": 1000,
-                "name": "Jaime Soriano Pastor"
+                "cmdline": "/tmp/go-build661675341/b001/socket.test -data"
             }
         }
+    },
+    "user": {
+        "full_name": "Jaime Soriano Pastor",
+        "id": 1000
```
Comment (webmat) [inline metricbeat/module/system/socket/_meta/data.json]: Almost missed this one. In ECS, IDs are always keyword. Windows uids are alphanumeric, for example.

Please leave this as a string.

File: dev-tools/ecs-migration.yml [hunk 4]
```diff
@@ -1211,6 +1211,43 @@
 
 ## Modules
 
+### System
+
+- from: system.socket.direction
```
Comment (ruflin) [inline dev-tools/ecs-migration.yml]: @simianhacker Could you see that and of the fields change in this file could have an effect on Infra UI?

File: metricbeat/helper/socket/listeners.go (hunk 1 above)
Comment (ruflin) [reply metricbeat/helper/socket/listeners.go]: If we keep the both values around, we can't easily disable it when `migration.enabled: true` is set. It's possible but would require quite a bit of work. Because of this I suggest we make this a breaking change and don't keep the old version around.

Comment (ruflin) [reply metricbeat/helper/socket/listeners.go]: @andrewkroh One more place where we could use the constants from ECS in Go code ;-)

File: metricbeat/module/system/socket/_meta/data.json [hunk 5]
```diff
@@ -12,27 +12,40 @@
...
+        "args": [
+            "/tmp/go-build661675341/b001/socket.test",
+            "-data"
+        ],
+        "executable": "/tmp/go-build661675341/b001/socket.test",
+        "name": "socket.test",
+        "pid": 9341
+    },
+    "server": {
+        "ip": "::",
```
Comment (ruflin) [inline metricbeat/module/system/socket/_meta/data.json]: Does ES treat this as a valid ip?

File: metricbeat/module/system/socket/_meta/data.json [hunk 6]
```diff
@@ -12,27 +12,40 @@
...
             "process": {
-                "cmdline": "/tmp/go-build827888631/b001/socket.test -data",
-                "command": "socket.test",
-                "exe": "/tmp/go-build827888631/b001/socket.test",
-                "pid": 19047
-            },
-            "user": {
-                "id": 1000,
-                "name": "Jaime Soriano Pastor"
+                "cmdline": "/tmp/go-build661675341/b001/socket.test -data"
```
[the remark this answers is not in the document]
Comment (ruflin) [reply metricbeat/module/system/socket/_meta/data.json]: It's kind of odd to keep this here. Could we come up with a name that is not in ECS but still put it under the top level process?

File: metricbeat/helper/socket/listeners.go (hunk 1 above)
Comment (jsoriano) [reply metricbeat/helper/socket/listeners.go]: Ok, I'll change this and add a note about the breaking change.

File: metricbeat/module/system/socket/socket.go [hunk 7]
```diff
@@ -258,54 +263,107 @@ func newConnection(diag *linux.InetDiagMsg) *connection {
...
 
 	if c.Username != "" {
-		evt.Put("user.name", c.Username)
+		fields.Put("user.full_name", c.Username)
 	}
 
 	if c.ProcessError != nil {
-		evt.Put("process.error", c.ProcessError.Error())
+		// XXX, would this be the proper place for a process error code?
+		fields.Put("error.code", c.ProcessError.Error())
```
[the remark this answers is not in the document]
Comment (jsoriano) [reply metricbeat/module/system/socket/socket.go]: Could `error` be nested? for this case it could make sense to leave it on `process.error.code`.

File: metricbeat/module/system/socket/socket.go [hunk 8]
```diff
@@ -258,54 +263,107 @@ func newConnection(diag *linux.InetDiagMsg) *connection {
...
 		} else if c.PID == 0 {
 			process["command"] = "kernel"
 		}
+
+		fields["process"] = process
 	}
 
 	if c.RemotePort != 0 {
+		// Aliases for this are not going to be possible, keeping
+		// duplicated fields by now for backwards comatibility
```
[the remark this answers is not in the document]
Comment (jsoriano) [reply metricbeat/module/system/socket/socket.go]: I think we should keep BC at least for a time if we cannot offer an alternative with aliases.

File: metricbeat/module/system/socket/socket.go (hunk 2 above)
Comment (jsoriano) [reply metricbeat/module/system/socket/socket.go]: For `listening` I think we can be quite sure that this is a server, but for incoming/outgoing connections I am not sure we can assume this is a client-server protocol, source/destination doesn't assumes protocols.

(Actually local/remote is the terminology used on Linux for this.)

File: metricbeat/mb/testing/data_generator.go [hunk 9]
```diff
@@ -77,29 +77,25 @@ func WriteEventsCond(f mb.EventsFetcher, t testing.TB, cond func(e common.MapStr
 		return fmt.Errorf("no events were generated")
```
Comment (jsoriano) [inline metricbeat/mb/testing/data_generator.go]: Changes in this file moved to its own PR #10367 10367

File: metricbeat/module/system/socket/socket.go (hunk 2 above)
Comment (webmat) [reply metricbeat/module/system/socket/socket.go]: Yes, right now ECS defines the client/server and source/destination pairs. I think there's some contexts where local/remote would also be clearer.

If you're not comfortable using client/server in some of these situations, can you use source/destination? Note that ECS assumes that event sources fill in source/destination all of the time, and client/server only when it makes sense.

So in all the places where you have client/server, you should also duplicate these field sets to source/destination. src/dest are considered the "baseline" that you can always rely on being there.

File: metricbeat/module/system/socket/_meta/fields.yml [hunk 10]
```diff
@@ -63,30 +60,31 @@
...
+      type: alias
+      path: user.id
+      migration: true
 
     - name: user.name
-      type: keyword
-      description: >
-        Name of the user running the process.
+      type: alias
+      path: user.full_name
```
[the remark this answers is not in the document]
Comment (jsoriano) [reply metricbeat/module/system/socket/_meta/fields.yml]: Ok, both name and full name collected now.

File: metricbeat/module/system/socket/socket.go (hunk 2 above)
Comment (jsoriano) [reply metricbeat/module/system/socket/socket.go]: Yes, using source/destination now, and server only for listening sockets.

I still keep system.socket.local/remote becase we'd need N-N aliases for migration.

File: metricbeat/module/system/socket/socket.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (jsoriano) [reply metricbeat/module/system/socket/socket.go]: Ok, keeping this in `error.code` then.

File: metricbeat/module/system/socket/socket.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (jsoriano) [reply metricbeat/module/system/socket/socket.go]: Btw, one advantage of local/remote vs source/destination is that you know that local refers to local resources. You can also obtain this info from source/destination, but you need to check direction to know which one is the local one.

**Comment by jsoriano** [comment]
Ouch :man_facepalming: variable names in test fixed.

**Comment by jsoriano** [comment]
Waiting for #10516 to update and retest this

**Comment by jsoriano** [comment]
CI failures not related, merging.
